src/mlpack/tests/main_tests/perceptron_test.cpp

```diff
@@ -0,0 +1,200 @@
+{
+  arma::mat inputData;
+  if (!data::Load("trainSet.csv", inputData))
+    BOOST_FAIL("Cannot load train dataset trainSet.csv!");
+
+  // Get the labels out.
+  arma::Row<size_t> labels(inputData.n_cols);
+  for (size_t i = 0; i < inputData.n_cols; ++i)
+    labels[i] = inputData(inputData.n_rows - 1, i);
+
+  // Delete the last row containing labels from input dataset.
+  inputData.shed_row(inputData.n_rows - 1);
+
+  arma::mat testData;
+  if (!data::Load("testSet.csv", testData))
+    BOOST_FAIL("Cannot load test dataset testSet.csv!");
+
+  // Delete the last row containing labels from test dataset.
+  testData.shed_row(testData.n_rows - 1);
+
+  size_t testSize = testData.n_cols;
+
+  // Input training data.
+  SetInputParam("training", std::move(inputData));
+  SetInputParam("labels", std::move(labels));
+
+  // Input test data.
+  SetInputParam("test", std::move(testData));
+
+  mlpackMain();
+
+  // Check that number of output points are equal to number of input points.
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::Row<size_t>>("output").n_cols,
+                      testSize);
+
+  // Check output have only single row.
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::Row<size_t>>("output").n_rows, 1);
+}
+
+/**
+ * Check that last row of input file is used as labels
+ * when labels are not passed specifically.
+ */
+BOOST_AUTO_TEST_CASE(PerceptronLabelsLessDimensionTest)
+{
+  arma::mat inputData;
+  if (!data::Load("trainSet.csv", inputData))
+    BOOST_FAIL("Cannot load train dataset trainSet.csv!");
+
+  arma::mat testData;
+  if (!data::Load("testSet.csv", testData))
+    BOOST_FAIL("Cannot load test dataset testSet.csv!");
+
+  // Delete the last row containing labels from test dataset.
+  testData.shed_row(testData.n_rows - 1);
+
+  size_t testSize = testData.n_cols;
+
+  // Input training data.
+  SetInputParam("training", std::move(inputData));
+
+  // Input test data.
+  SetInputParam("test", std::move(testData));
+
+  mlpackMain();
+
+  // Check that number of output points are equal to number of input points.
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::Row<size_t>>("output").n_cols,
+                      testSize);
+
+  // Check output have only single row.
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::Row<size_t>>("output").n_rows, 1);
+}
+
+/**
+ * Ensure that saved model can be used again.
+ */
+BOOST_AUTO_TEST_CASE(PerceptronModelReuseTest)
+{
+  arma::mat inputData;
+  if (!data::Load("trainSet.csv", inputData))
+    BOOST_FAIL("Cannot load train dataset trainSet.csv!");
+
+  arma::mat testData;
+  if (!data::Load("testSet.csv", testData))
+    BOOST_FAIL("Cannot load test dataset testSet.csv!");
+
+  // Delete the last row containing labels from test dataset.
+  testData.shed_row(testData.n_rows - 1);
+
+  size_t testSize = testData.n_cols;
+
+  // Input training data.
+  SetInputParam("training", std::move(inputData));
+
+  // Input test data.
+  SetInputParam("test", std::move(testData));
```

Comment:
You could just make a copy here, and then you wouldn't need to reload the test set later.  I think this is also true for some other tests you've written in this PR.

Reply:
Yup! that's a better way around :)

Reply:
Ah, I meant something a little different---you could do 

```
SetInputParam("test", testData);
```

here, and then later

```
SetInputParam("test", std::move(testData));
```

The first call to `SetInputParam()` will make a copy implicitly.

Reply:
Sorry, I misunderstood. Will correct it shortly.

Reply:
This pattern appears lots of other places in this PR too, so if you can handle it everywhere, I'd really appreciate it.  Thanks! :)